src/lxc/lxccontainer.c

```diff
@@ -4659,6 +4680,158 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
 
 WRAP_API_2(bool, lxcapi_restore, char *, bool)
 
+static int do_lxcapi_mount(struct lxc_container *c,
+			const char *source, const char *target,
+			const char *filesystemtype, unsigned long mountflags,
+			const void *data, struct lxc_mount *mnt) {
+	int fd;
+	char *template, *suff, *path;
+	pid_t pid, init_pid;
+	size_t len;
+	struct stat sb;
+	int ret = -1;
+
+	if (!c || !c->lxc_conf) {
+		ERROR("Container or configuration is NULL");
+		goto out_err;
+	}
+
+	len = strlen(c->lxc_conf->lxc_shmount.path_host) + sizeof("/lxcmountXXXXXX");
+	template = malloc(len);
+	if (!template) {
+		SYSERROR("Not enough memory");
+		goto out_err_free;
+	}
+	ret = snprintf(template, len, "%s/lxcmountXXXXXX", c->lxc_conf->lxc_shmount.path_host);
+	if (ret < 0 || (size_t)ret >= len) {
+		SYSERROR("Error writing shmounts tempdir name");
+		goto out_err_free;
+	}
+
+	/* Create a temporary dir / file
+	 * under the shared mountpoint
+	 */
+	ret = stat(source, &sb);
+	if (ret < 0) {
+		SYSERROR("Could not read source information");
+		goto out_err_free;
+	}
+
+	if (S_ISREG(sb.st_mode)) {
+		fd = mkstemp(template);
+		if (fd < 0) {
+			SYSERROR("Could not create shmounts temporary file");
+			goto out_err_free;
+		}
+	} else if (S_ISDIR(sb.st_mode)) {
+		template = mkdtemp(template);
+		if (!template) {
+			SYSERROR("Could not create shmounts temporary dir");
+			goto out_err_free;
+		}
+	} else {
+		ERROR("Unexpected type of shared mountpoint");
+		goto out_err_free;
+	}
+
+	/* Do the fork */
+	pid = fork();
+	if (pid < 0) {
+		SYSERROR("Could not fork");
+		goto out_err_free;
+	}
+
+	if (pid == 0) {
+		int exit_status = EXIT_FAILURE;
+
+		/* Do the mount */
+		ret = mount(source, template, filesystemtype, mountflags, data);
+		if (ret < 0) {
+			SYSERROR("Failed to mount \"%s\" onto \"%s\"", source, template);
+			goto out_child_err_temp;
+		}
+
+		init_pid = do_lxcapi_init_pid(c);
+		if (init_pid < 0) {
+			ERROR("Failed to obtain container's init pid");
+			goto out_child_err_temp;
+		}
+
+		/* Enter the container namespaces */
+		if (am_guest_unpriv() && !lxc_list_empty(&c->lxc_conf->id_map)) {
+			/* if fully unprivileged */
+			ret = setns(init_pid, CLONE_NEWUSER);
+			if (ret < 0) {
+				SYSERROR("Failed to join container's userns");
+				goto out_child_err_temp;
```

Comment:
`_exit(EXIT_FAILURE)` here directly since there's nothing to cleanup yet.

Reply:
So about those `setns()` what you're doing now is invalid. If you do

    man setns

you'll see that the first argument of `setns()` is `int fd` which is a file descriptor but you're passing the `init_pid`. I remember that I jotted it down this way in the spec we wrote so I should've clarified this (and please update the spec accordingly :)). What you need to do is:

    int userns_fd = open("/proc/<init_pid>/ns/user", O_RDONLY | O_CLOEXEC);
    int mntns_fd = open("/proc/<init_pid>/ns/mnt", O_RDONLY | O_CLOEXEC);

and pass the correct `fd` to the corresponding `setns()`. Which also means you need to cleanup those `fd`s too later on. ;)

The good news about this is that it won't get more complex because we have a helper in `utils.{c,h}` that does all of this for you:

```
bool switch_to_ns(pid_t pid, const char *ns) {
	int fd, ret;
	char nspath[MAXPATHLEN];

	/* Switch to new ns */
	ret = snprintf(nspath, MAXPATHLEN, "/proc/%d/ns/%s", pid, ns);
	if (ret < 0 || ret >= MAXPATHLEN)
		return false;

	fd = open(nspath, O_RDONLY);
	if (fd < 0) {
		SYSERROR("failed to open %s", nspath);
		return false;
	}

	ret = setns(fd, 0);
	if (ret) {
		SYSERROR("failed to set process %d to %s of %d.", pid, ns, fd);
		close(fd);
		return false;
	}
	close(fd);
	return true;
}
```